Video analyser

aquatiko opened the pull request:
The main aim of this app is to analyze(i.e do OCR on) videos taken at high fps.

## Current features
- Allows users to select any video from thier storage.
- Load the video in backend and provide an alalysis window giving access to each video frame.
- Do OCR on frames in that window.

## ToDo
- [x] Rotation parameter of videos required for OCR is hardcoded for now because avilable rotation metadata is giving weird/incorrect OCR.
- [x] Total number of frames read at a time is hardcoded for now as accessing a frame id viz. not present gives meomory read errors.
- [x] Able to do OCR on 100 frames only in a session. Find the reason and fix this constraint.

## Main files of interest
From videoAnalyser
[MainActivity.java](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-ce9ab72c6602e37de118a7629012e9f1)
[activity_main.xml](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-690c15ceb6336040cf26ec505cbcc93b)

From Camera2SlowMotion
[CameraFragment.kt](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-688168251bd22bedafbb200c80e769da)
[PermissionsFragment.kt](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-1b4fd01602ea4ed7af651019f699706e)
[SelectorFragment.kt](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-b41b6c60c516a1b90227e696269bd990)
[activity_camera.xml](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-72c9bd32246124fa50c8c26254f48549)
[fragment_camera.xml](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-af2f711f3d15eacfd7d69001bf93651b)

## Future steps
Integrate this with [Camera2SlowMotion](https://github.com/android/camera-samples/tree/master/Camera2SlowMotion) and replace video selection from storage with on demand video capture and processing.

aquatiko commented:
> Rotation parameter of videos required for OCR is hardcoded for now because avilable rotation metadata is giving weird/incorrect OCR.

I took 4 videos at 4 possible rotation of 0, 90, 180, 270 degree and when they are stored in SD card, they are automatically reoriented at 0 degree and there is no need of handling rotation later in the analyser part.
The OCR can be given a hardcoded rotation of 0 degree for the OCR to work.

aquatiko commented:
> Able to do OCR on 100 frames only in a session. Find the reason and fix this constraint.

This was a flake error. Working fine now.

aquatiko commented on videoanalyser/app/src/main/java/com/example/videoanalyser/MainActivity.java:
```diff
@@ -81,14 +82,12 @@ public void onClick(View view) {
-                rotationOfVideo = mediaMetadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_ROTATION);
-                List<Bitmap> frameList = mediaMetadataRetriever.getFramesAtIndex(100, 100);
+                totalFrames = Integer.valueOf(mediaMetadataRetriever.extractMetadata(METADATA_KEY_VIDEO_FRAME_COUNT));
+                List<Bitmap> frameList = mediaMetadataRetriever.getFramesAtIndex(0, Math.max(0, totalFrames-safeFrames));
```
`safeFrames` are required here because while recording a video at high fps, there are cases when the last frames are not taken correctly.
Eg: Consider a video of time 4.606 seconds taken at 120fps, there will be 553 frames as suggested in metadata.
But when we do the math, we find that there should be 552.72 frames and the last frames is not correctly sotred.

To account this, `safeFrames` value trims some frames from the end.

patrickmeiring replied:
Thanks for the explanation. From your explanation, it sounds like safeFrames = 1 should work (unless a bunch of frames are stored together due to compression?) but I'm not too worried about this since I understand we're trying to do the analysis without storing the video at all.

aquatiko replied:
Yeah, I also thought `safeFrames = 1` should be enough but for some `safeFrames = 2` was required else I was getting memory leaks. So just to be on safe side I added it as `10` for now.
Thinking now, I realise that this could be the reason.
>(unless a bunch of frames are stored together due to compression?)

patrickmeiring commented on videoanalyser/app/src/main/java/com/example/videoanalyser/MainActivity.java:
```diff
@@ -0,0 +1,132 @@
+        });
+
+        analyzeBtn.setOnClickListener(new View.OnClickListener() {
+            @RequiresApi(api = Build.VERSION_CODES.P)
```
Nit: Just curious, where does this come from/why do we need this annotation?

aquatiko replied:
It was compiler suggested. Tells programmer that an android Pie(9) API is needed.
I think that would limit the number of devices, I will look into this more.

aquatiko replied:
The function `getFramesAtIndex` at Line 86 reuires this API level.

`List<Bitmap> frameList = mediaMetadataRetriever.getFramesAtIndex(0, Math.max(0, totalFrames-safeFrames))`

patrickmeiring commented on videoanalyser/app/src/main/java/com/example/videoanalyser/MainActivity.java:
```diff
@@ -0,0 +1,132 @@
+    private Intent fileIntent;
+    private MediaMetadataRetriever mediaMetadataRetriever;
+    private Uri fileUri;
+    private  Integer totalFrames=0;
```
Nit: Please run the formatter over this file. I think this should be "private Integer totalFrames = 0".
Can see some strange whitespace issues further down the file (extra lines before end of functions/anonymous class declarations, odd indenting, etc.).

aquatiko replied:
Sure, thanks for pointing!

patrickmeiring commented on videoanalyser/app/src/main/java/com/example/videoanalyser/MainActivity.java:
```diff
@@ -0,0 +1,132 @@
+                    // analysis will be done here (forking in a separate thread)
+                    imageHolder = InputImage.fromBitmap(frameList.get(i), 0);
+                    final int finalI = i;
+                    Task<Text> result =
```
Observation: Right now this is an experiment, but when we get to implementing this for real, I think it makes sense to track/manage the number of asynchronous tasks you have going on at once to avoid exhausting system resources. This may occur naturally in a binary search implementation anyway, as the next frame to analyse depends on the result of the earlier frame.

aquatiko replied:
Interesting idea! I will patch that

mwarton replied:
We will also need a way to get the results back into the main thread.

Most likely have worker threads with a queue of tasks in one direction and results in the other direction.

patrickmeiring commented on videoanalyser/app/src/main/java/com/example/videoanalyser/MainActivity.java:
```diff
@@ -0,0 +1,132 @@
+            public void onClick(View view) {
+                fileIntent = new Intent(Intent.ACTION_GET_CONTENT);
+                fileIntent.setType("*/*");
+                startActivityForResult(fileIntent, 10);
```
Please define a constant for '10', so we know it is the same value we are switching on down in 'onActivityResult'.

aquatiko replied:
Done in refactoring commit.

patrickmeiring commented on videoanalyser/app/src/main/java/com/example/videoanalyser/MainActivity.java:
```diff
@@ -0,0 +1,132 @@
+        });
+    }
+
+    @SuppressLint("MissingSuperCall")
```
nit: Any reason we don't also call into the superclass onActivityResult method here?
super.onActivityResult(requestCode, resultCode, data)

aquatiko replied:
It seems like I accidently deleted it and added a supress warning on compiler suggestion. Added in refactoring commit.

patrickmeiring reviewed (commented):
Thanks for committing, Rohit!

Just a few nits, but otherwise LGTM. Look forward to seeing the analysis integrated with camera capture!

mwarton commented on videoanalyser/app/src/main/java/com/example/videoanalyser/MainActivity.java:
```diff
@@ -0,0 +1,134 @@
+                                                @Override
+                                                public void onFailure(@NonNull Exception e) {
+                                                    // Task failed with an exception
+                                                    // ...
```
Please at least log the exception, nothing worse than code failing silently and driving youself mad trying to find what the problem is.

aquatiko replied:
Done in refactoring commit.

mwarton commented on videoanalyser/app/src/main/java/com/example/videoanalyser/MainActivity.java:
```diff
@@ -0,0 +1,134 @@
+                                            // Task completed successfully
+                                            // ...
+                                            results.add(visionText.getText());
+                                            Log.d("Rokus logs", "Text detected at index:" + finalI + " " + visionText.getText());
```
The level of indentation here is getting a bit extreme, can you maybe move the bulk of this code to an analyze method on the main class and just have the onClick handler call that method?

aquatiko replied:
Done in refactoring commit.

aquatiko committed:
videoAnalyser core functionality added

aquatiko commented:
I integrated everything till `refactoring` commit into [Camera2SlowMotion](https://github.com/android/camera-samples/tree/master/Camera2SlowMotion).

The files of interest are:

From videoAnalyser
[MainActivity.java](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-ce9ab72c6602e37de118a7629012e9f1)
[activity_main.xml](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-690c15ceb6336040cf26ec505cbcc93b)

From Camera2SlowMotion
[CameraFragment.kt](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-688168251bd22bedafbb200c80e769da)
[PermissionsFragment.kt](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-1b4fd01602ea4ed7af651019f699706e)
[SelectorFragment.kt](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-b41b6c60c516a1b90227e696269bd990)
[activity_camera.xml](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-72c9bd32246124fa50c8c26254f48549)
[fragment_camera.xml](https://github.com/googleinterns/screen-latency-testing/pull/5/files#diff-af2f711f3d15eacfd7d69001bf93651b)

patrickmeiring commented:
@aquatiko It looks like you have committed a bunch of IntelliJ caches and generated files. Note that we do not want to track these in source control as they clutter the repository (please only include source code, build scripts and basic project config). Can you amend your last commit to remove the files we don't want and add a .gitignore file that ignores these files for future (like we did in #1)? :)

I will take a look at your second commit when you have done this (the review pane is currently stalling for me with 1700 files).

mwarton commented on videoanalyser/app/src/main/java/com/example/videoanalyser/MainActivity.java:
```diff
@@ -57,14 +58,16 @@ protected void onCreate(Bundle savedInstanceState) {
                 fileIntent = new Intent(Intent.ACTION_GET_CONTENT);
                 fileIntent.setType("*/*");
-                startActivityForResult(fileIntent, 10);
+                startActivityForResult(fileIntent, filePickerRequestCode);
```
Can we name this FILE_PICKER_REQUEST_CODE to make it clear that it is a constant?

patrickmeiring replied:
(Styleguide reference: https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names )

patrickmeiring replied:
^ Assume that is the styleguide we want you to follow. And if we make comments inconsistent with that guide, feel free to push back :)

aquatiko commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/fragments/CameraFragment.kt:
```diff
@@ -0,0 +1,488 @@
+                    MediaScannerConnection.scanFile(
+                            view.context, arrayOf(outputFile.absolutePath), null, null)
+                    Log.d("Rokus Logs:", "reached intent call in kotlin")
+                    // Launch external activity via intent to play video recorded using our provider
```
This functionality from the boilerplate was suppressed and I added the following code to connect both the parts. i.e by using an `intent` to another activity.

patrickmeiring replied:
Noted, thanks. For review and code history purposes, it would be nice if you could separate your changes from the base sample app.

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/MainActivity.java:
```diff
@@ -107,24 +107,24 @@ private void analyze(){
+                                        public void onFailure(@NonNull Exception e) {
+                                            // Task failed with an exception
+                                            e.printStackTrace();
+                                            e.getCause();
```
Left over from debugging? Throwable.getCause() has no side-effects and the result is unused, so it makes no sense to call it here.

patrickmeiring replied:
nit: I don't think this file should be changed in 6da0c5 (since it is the import of sample code), if you must change the indentation maybe do it in the later commit (or fix the earlier commit). Also, I think the comment about e.getCause() is still outstanding.

patrickmeiring commented on Camera2SlowMotion/utils/src/main/java/com/example/android/camera/utils/AutoFitSurfaceView.kt:
```diff
@@ -0,0 +1,79 @@
+        if (aspectRatio == 0f) {
+            setMeasuredDimension(width, height)
+        } else {
+
```
nit: Surplus whitespace. (May have been like this in the sample code already.)

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/fragments/CameraFragment.kt:
```diff
@@ -0,0 +1,488 @@
+//                        data = FileProvider.getUriForFile(view.context, authority, outputFile)
+//                        flags = Intent.FLAG_GRANT_READ_URI_PERMISSION or
+//                                Intent.FLAG_ACTIVITY_CLEAR_TOP
+//                    })
```
Prefer to delete rather than comment out -- if you separate out the base app from your changes, source control will track the deletion and we will always be able to recover the code if we need it by going through the history.

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/fragments/CameraFragment.kt:
```diff
@@ -0,0 +1,488 @@
+                    }
+
+
+
```
nit: Extra whitespace here, can you clean this up?

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/fragments/CameraFragment.kt:
```diff
@@ -0,0 +1,488 @@
+                }
+            }
+
+            true
```
Don't think this should be here. As above -- might be easier for you and me to pick little things like this up if you can separate your deltas from the base sample app.

patrickmeiring replied:
Still seeing this in revision 6da0c5. Looks like this was actually in the sample code: https://github.com/android/camera-samples/blob/master/Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/fragments/CameraFragment.kt#L379
Can you clean it up in the commit with your changes that follows? :)

patrickmeiring reviewed (commented):
Thanks Rohit :) I've made a few comments. Is it possible for you to separate your changes from the base/sample app?

aquatiko committed:
Importing Camera2SlowMotion sample app

aquatiko committed:
Migrating videoAnalyser into Camera2SlowMotion

Deleting project files of VideoAnalyser and adding relevant parts into
Camera2SlowMotion sample app

aquatiko committed:
interpolate logic for videoAnalyser

Involves intent calls between kotlin(Camera2SlowMotion) and
java(videoAnalyser)

aquatiko committed:
videoAnalyser starter code

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/MainActivity.java:
```diff
@@ -0,0 +1,137 @@
+                                        @Override
+                                        public void onFailure(@NonNull Exception e) {
+                                            // Task failed with an exception
+
```
nit: Random blank line here

patrickmeiring commented on videoanalyser/app/src/main/java/com/example/videoanalyser/MainActivity.java:
```diff
@@ -1,14 +1,132 @@
+                    analyze();
+                }catch (Exception e){
+                    e.printStackTrace();
+                    e.getCause();
```
nit: Missed deleting an e.getCause() here. Like the other one you removed, I don't think this there is any point to calling this method as it has no side effects and the result is not used.

patrickmeiring commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/MainActivity.java:
```diff
@@ -0,0 +1,137 @@
+        totalFrames = Integer.valueOf(mediaMetadataRetriever.extractMetadata(METADATA_KEY_VIDEO_FRAME_COUNT));
+        List<Bitmap> frameList = mediaMetadataRetriever.getFramesAtIndex(0, Math.max(0, totalFrames-safeFrames));
+
+        for(int i=0; i<frameList.size();i++) {
```
nit: Still seeing some strange formatting here and elsewhere. Can you run the formatter over this and the other files you changed?

mwarton replied:
can you set up either format on save or format on commit?

patrickmeiring commented:
Looking a lot better! Great work :)

LGTM % nits. Just a few issues that were pointed out earlier (formatting, e.getCause()) that still need to be addressed everywhere.

patrickmeiring reviewed (commented):
Looking a lot better! Great work :)

LGTM % nits. Just a few issues that were pointed out earlier (formatting, e.getCause()) that still need to be addressed everywhere.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/MainActivity.java:
```diff
@@ -38,7 +38,7 @@
 public class MainActivity extends AppCompatActivity {
     private Integer totalFrames=0;
```
I think these two should be final and CONSTANT_CASE two right?

aquatiko replied:
`safeFrames` can be constant but I'm setting `totalFrames` from metadata later on. Initialized here as I didn't wanted a garbage value to indicate that `totalFrames` is set.

mwarton commented on Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/fragments/CameraFragment.kt:
```diff
@@ -359,16 +357,21 @@ class CameraFragment : Fragment() {
+                    intent.putExtra("file URI", FileProvider.getUriForFile(view.context, authority, outputFile).toString())
+                    Log.d("Rokus Logs:", "starting intent call in kotlin")
+                    try {
+                        // some code
```
delete useless comments?

aquatiko committed:
add missing layout of activity_main.xml

aquatiko committed:
added on device result viewing

aquatiko committed:
remove leftover files of videoAnalyser

aquatiko committed:
applied google java style formatter

mwarton reviewed (approved):
LGTM
